### PR TITLE
unison: fix build with ocamlPackages_4_09

### LINKED ISSUE #61867: unison fails t build, citing 'Unbound value Gpointer.region_of_string' (nixOs 19.03)
## Issue description

Trying to install unison (2.51.2) on nixos 19.03 fails thusly:

```
ocamlopt -I /nix/store/mk7a5hkrhi24yhzi903cd88m9jjcjzmc-lablgtk-2.18.6/lib/ocaml/4.06.1/site-lib/lablgtk2 -g -unsafe-string -I lwt -I ubase -I system -I fsmonitor -I fsmonitor/linux -I fsmonitor/windows -I system/generic -I lwt/generic -I +lablgtk2 -c /build/source/src/uigtk2.ml
File "/build/source/src/uigtk2.ml", line 26, characters 24-41:
Warning 3: deprecated: String.capitalize
Use String.capitalize_ascii instead.
File "/build/source/src/uigtk2.ml", line 97, characters 5-30:
Error: Unbound value Gpointer.region_of_string

```

### Steps to reproduce

```
nix-env -i unison
```

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.41, NixOS, 19.09.git.f29d398 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - nixpkgs: `/nix/var/nixpkgs/release-19.03.f29d398/default.nix`
```

### PR DESCRIPTION
**Pull request by winterdeaf**
###### Motivation for this change
Fixes #61867 and #61505, bumps the ocaml version unison is built
against to 4.09. The patches included here appear in the trunk version
of unison, but were not backported to 2.51.2.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vbgl** [comment]
Why choosing OCaml 4.09 rather than the default version of OCaml in `nixpkgs`?

**Comment by m-scr** [comment]
Local and remote instances need to be built with the same (or very close) Ocaml versions to communicate -- it could be good to pin it to a fixed version, but I don't know whether 4.09 is ideal. Debian/Ubuntu are shipping Unison 2.48 w/ Ocaml 4.08, Archlinux is using 4.09. The current default version in `nixpkg` is 4.07, which I cannot sync with my other hosts.

We could maybe add an configuration option to specify the Ocaml to use, if there is an easy way to make it select the correct patches.

**Comment by vbgl** [comment]
Generally speaking, packages in `nixpkgs` can easily be overridden. For instance, here is how to change the version of OCaml used for building `unison`:

~~~
unison.override { ocamlPackages = ocaml-ng.ocamlPackages_4_02; }
~~~

The patches that you add in this PR fix the build of unison with all versions of OCaml between 4.06 and 4.09 (included). Therefore, I don’t see a good reason to choose a different default version of OCaml than in the rest of `nixpkgs`.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -22044,7 +22044,7 @@ in
 
   unison = callPackage ../applications/networking/sync/unison {
     enableX11 = config.unison.enableX11 or true;
-    ocamlPackages = ocaml-ng.ocamlPackages_4_05;
+    ocamlPackages = ocaml-ng.ocamlPackages_4_09;
```
Comment (vbgl) [inline pkgs/top-level/all-packages.nix]:
```suggestion
```

**Review by vbgl — CHANGES_REQUESTED** [review]

**Comment by m-scr** [comment]
> ```
> unison.override { ocamlPackages = ocaml-ng.ocamlPackages_4_02; }
> ```

Sounds good to me.

**Comment by vbgl** [comment]
@GrahamcOfBorg build unison

**Review by vbgl — APPROVED** [review]

**Comment by vbgl** [comment]
Merged into 20.03 as 27eca645524330dd46f10387538ada32206c8b02

Thanks!

**Pull request merged**